src/backend/catalog/catalog.c

```diff
@@ -558,10 +625,31 @@ GetNewRelFileNode(Oid reltablespace, Relation pg_class, char relpersistence)
 
 		/* Generate the OID */
 		if (pg_class)
-			rnode.node.relNode = GetNewOidWithIndex(pg_class, ClassOidIndexId,
+		{
+			if (relpersistence == RELPERSISTENCE_TEMP && sql_dialect == SQL_DIALECT_TSQL)
+			{
+				/* Temp tables use temp OID logic */
+				rnode.node.relNode = GetNewTempOidWithIndex(pg_class, ClassOidIndexId,
 													Anum_pg_class_oid);
+			}
+			else
+			{
+				rnode.node.relNode = GetNewOidWithIndex(pg_class, ClassOidIndexId,
+													Anum_pg_class_oid);
+			}
+		}	
 		else
-			rnode.node.relNode = GetNewObjectId();
+		{
+			if (relpersistence == RELPERSISTENCE_TEMP && sql_dialect == SQL_DIALECT_TSQL)
+			{
+				/* Temp tables use temp OID logic */
+				rnode.node.relNode = GetNewTempObjectId();
+			}
+			else
+			{
+				rnode.node.relNode = GetNewObjectId();
+			}
+		}
```

Comment:
The code will look less invasive if we move most of the new logic generating OID into hook functions

Reply:
Can this be done?

Reply:
Done